sandbox/update-local-registry.py

```diff
@@ -0,0 +1,243 @@
+
+
+def build_image(name: str, args: argparse.Namespace) -> None:
+    info = image_info(name)
+    tmpdir = tempfile.mkdtemp()
+    try:
+        repo = info.repo or name
+        url = repo_url(repo)
+        branch = info.branch
+        logging.info('Checking out %s branch %s to build %s', url, branch, repo)
+        # GIT_TERMINAL_PROMPT=0 prevents Git from asking for credentials,
+        # which could happen if a typo meant the repository did not exist.
+        git_clone(tmpdir, url, branch)
+        git_rev = subprocess.run(
+            ['git', 'rev-parse', 'HEAD'],
+            stdin=subprocess.DEVNULL, stdout=subprocess.PIPE, check=True, cwd=tmpdir,
+            encoding='utf-8').stdout.strip()
+
+        downstream_image = downstream_path(name, args)
+        logging.info('Building image %s', downstream_image)
+        workdir = os.path.join(tmpdir, name) if info.repo else tmpdir
+        if os.path.exists(os.path.join(workdir, 'scripts', 'docker_build.sh')):
+            docker_build = ['scripts/docker_build.sh']
+        else:
+            docker_build = ['docker', 'build']
+        subprocess.run(
+            docker_build + [
+                '--label=org.label-schema.schema-version=1.0',
+                f'--label=org.label-schema.vcs-ref={git_rev}',
+                f'--label=org.label-schema.vcs-url={url}',
+                '--build-arg', f'KATSDPDOCKERBASE_REGISTRY={args.downstream}',
+                '--build-arg', f'KATSDPDOCKERBASE_MIRROR={args.http_mirror}',
+                '--pull=true',
+                '-t', downstream_image, '.'
+            ], stdin=subprocess.DEVNULL, check=True, cwd=workdir)
+        logging.info('Pushing %s', downstream_image)
+        docker_cmd('push', downstream_image)
+        logging.info('%s pushed', downstream_image)
+    finally:
+        shutil.rmtree(tmpdir)
+
+
+def copy_image(name: str, args: argparse.Namespace) -> None:
+    upstream_image = upstream_path(name, args)
+    downstream_image = downstream_path(name, args)
+    logging.info('Pulling %s', upstream_image)
+    docker_cmd('pull', upstream_image)
+    logging.info('Tagging %s -> %s', upstream_image, downstream_image)
+    docker_cmd('tag', upstream_image, downstream_image)
+    logging.info('Pushing %s', downstream_image)
+    docker_cmd('push', downstream_image)
+    logging.info('%s pushed', downstream_image)
+
+
+def tune_image(name: str, args: argparse.Namespace) -> None:
+    assert name.startswith('katsdpingest_')
+    lines = subprocess.run(
+        ['nvidia-smi', '--query-gpu=name,uuid', '--format=csv,noheader'],
+        stdout=subprocess.PIPE, stdin=subprocess.DEVNULL,
+        check=True, encoding='utf-8').stdout.splitlines()
+    reader = csv.reader(lines)
+    gpus: Dict[str, str] = {}        # Maps name to UUID
+    for row in reader:
+        gpus[row[0]] = row[1].strip()
+    if not gpus:
+        logging.warning('No NVIDIA GPUs detected - skipping autotuning for %s', name)
```

Comment:
This doesn't seem like it would be a temporary failure - shouldn't we bail at this point ?

Reply:
I'm undecided - it could still be useful if one was going to do a build on a machine with no GPU and then do just the autotuning elsewhere.

Maybe I should make this a failure, but sort the steps so that tuning happens last? That way even if you forget to specify `--exclude=tune` you'll get all the building done before it crashes out here.

Reply:
Thanks - for a script that is likely to be used outside of the office I think it is important not to bumble on too much and rather fail so that they don't think they have actually built it properly.